pyupgrade/_plugins/type_bases_object.py

```diff
@@ -0,0 +1,106 @@
+from __future__ import annotations
+
+import ast
+from typing import Iterable
+
+from tokenize_rt import Offset
+from tokenize_rt import Token
+
+from pyupgrade._ast_helpers import ast_to_offset
+from pyupgrade._data import register
+from pyupgrade._data import State
+from pyupgrade._data import TokenFunc
+from pyupgrade._token_helpers import find_closing_bracket
+from pyupgrade._token_helpers import find_open_paren
+from pyupgrade._token_helpers import remove_base_class
+
+
+def _should_move_right_edge(src: str) -> bool:
+    return src != ')'
+
+
+def _not_right_edge(token: Token) -> bool:
+    return token.src != ')' and token.name != 'NAME'
+
+
+def _last_part_function(i: int, src: str) -> int:
+    return i if src == ',' else i + 1
+
+
+def _remove_bases(tokens: list[Token], left: int, right: int) -> None:
+    del tokens[left + 1:right + 1]
+
+
+def _multiple_first(tokens: list[Token], left: int, right: int) -> None:
+    # only one base will be left, (tuple) -> (tuple,)
+    if sum(
+            1 if token.name == 'NAME' else 0
+            for token in tokens[left:find_closing_bracket(tokens, left)]
+    ) == 2:
+        tokens.insert(right + 1, Token('OP', ','))
+    # we should preserve the newline
+    if tokens[left + 1].name == 'NL':
+        # we should also preserve indents
+        if tokens[left + 2].name == 'UNIMPORTANT_WS':
+            del tokens[left + 3:right]
+        else:
+            del tokens[left + 2:right]
+    else:
+        del tokens[left + 1:right]
+
+
+def _multiple_last(tokens: list[Token], left: int, last_part: int) -> None:
+    type_call_open = find_open_paren(tokens, 0)
+    bases_open = find_open_paren(tokens, type_call_open + 1)
+    # only one base will be left, (tuple) -> (tuple,)
+    if sum(
+            1 if token.name == 'NAME' else 0
+            for token in tokens[bases_open:last_part + 1]
+    ) == 2:
+        del tokens[left + 1:last_part]
+    # we should preserve indents
+    elif tokens[last_part - 1].name == 'UNIMPORTANT_WS':
+        # we should also preserve the newline
+        if tokens[last_part - 2].name == 'NL':
+            del tokens[left:last_part - 2]
+        else:
+            del tokens[left:last_part - 1]
+    else:
+        del tokens[left:last_part]
+
+
+def remove_base_class_from_type_call(i: int, tokens: list[Token]) -> None:
+    remove_base_class(
+        i,
+        tokens,
+        should_move_right_edge=_should_move_right_edge,
+        not_right_edge=_not_right_edge,
+        last_part_function=_last_part_function,
+        do_brace_stack_pop_loop=False,
+        remove_bases=_remove_bases,
+        single_base_char=',',
+        multiple_first_char=')',
+        multiple_first=_multiple_first,
+        multiple_last=_multiple_last,
+    )
```

Comment:
this is a maintenance nightmare, I can't accept the code in this state

Reply:
well there is a significant difference between `A = type(...)` and `class A(...)` so even though most of the logic is the same, you can see for yourself how much differs(or at least how much differs in my implementation)
this is either something that can be solved by entirely separating the cases(which will cause code duplication) or by someone more experienced / more adapt at this sort of thing tackling the issue.

do you think this is something that I should keep working on, even though it may take a few more tries(if I even find a better solution with my current skill set), or should we close the PR here and leave this for someone more experienced to tackle?

Reply:
> well there is a significant difference between A = type(...) and class A(...)

then that should be a dead giveaway that they should **NOT** share code

Reply:
well that is what I thought at first but you requested that they do(unless I misunderstood your comment as to which function that is)

"this function should reuse the function call parsing function rather than duplicating its work"

Reply:
well I agreed with that, but it seems I have misunderstood a comment of yours, understanding that you wanted them to share code(said comment - "this function should reuse the function call parsing function rather than duplicating its work")
😞

Reply:
thank you and sorry for the misunderstanding, will use it :)